flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,381 @@
+	private final Configuration flinkClientConfig;
+
+	/** When ResourceManager failover, the max attempt should recover. */
+	private final AtomicLong currentMaxAttemptId = new AtomicLong(0);
+
+	private final AtomicLong currentMaxPodId = new AtomicLong(0);
+
+	private final String clusterId;
+
+	private FlinkKubeClient kubeClient;
+
+	/** The number of pods requested, but not yet granted. */
+	private int numPendingPodRequests;
+
+	public KubernetesResourceManager(
+		RpcService rpcService,
+		String resourceManagerEndpointId,
+		ResourceID resourceId,
+		Configuration flinkConfig,
+		HighAvailabilityServices highAvailabilityServices,
+		HeartbeatServices heartbeatServices,
+		SlotManager slotManager,
+		JobLeaderIdService jobLeaderIdService,
+		ClusterInformation clusterInformation,
+		FatalErrorHandler fatalErrorHandler,
+		ResourceManagerMetricGroup resourceManagerMetricGroup) {
+		super(
+			rpcService,
+			resourceManagerEndpointId,
+			resourceId,
+			highAvailabilityServices,
+			heartbeatServices,
+			slotManager,
+			jobLeaderIdService,
+			clusterInformation,
+			fatalErrorHandler,
+			resourceManagerMetricGroup);
+
+		this.flinkConfig = flinkConfig;
+		this.clusterId = flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID);
+
+		this.workerNodeMap = new ConcurrentHashMap<>();
+
+		this.numPendingPodRequests = 0;
+
+		this.numberOfTaskSlots = flinkConfig.getInteger(TaskManagerOptions.NUM_TASK_SLOTS);
+		this.taskExecutorResourceSpec = TaskExecutorResourceUtils.resourceSpecFromConfig(flinkConfig);
+		this.defaultTaskManagerMemoryMB = taskExecutorResourceSpec.getTotalProcessMemorySize().getMebiBytes();
+		this.defaultCpus = flinkConfig.getDouble(KubernetesConfigOptions.TASK_MANAGER_CPU, numberOfTaskSlots);
+
+		this.slotsPerWorker = createWorkerSlotProfiles(flinkConfig);
+
+		// Load the flink config uploaded by flink client
+		this.flinkClientConfig = GlobalConfiguration.loadConfiguration();
+	}
+
+	@Override
+	protected void initialize() throws ResourceManagerException {
+		LOG.info("Initializing Kubernetes client.");
+		LOG.info("KubernetesResourceManager.initialize clusterId:{}", clusterId);
+
+		this.kubeClient = createFlinkKubeClient();
+
+		try {
+			getWorkerNodesFromPreviousAttempts();
+		} catch (Exception e) {
+			throw new ResourceManagerException(e);
+		}
+
+		this.kubeClient.watchPodsAndDoCallback(getTaskManagerLabels(), this);
+	}
+
+	@Override
+	public CompletableFuture<Void> onStop() {
+		// shut down all components
+		Throwable firstException = null;
+
+		if (kubeClient != null) {
+			try {
+				kubeClient.close();
+			} catch (Throwable t) {
+				firstException = t;
+			}
+		}
+
+		final CompletableFuture<Void> terminationFuture = super.onStop();
+
+		if (firstException != null) {
+			return FutureUtils.completedExceptionally(new FlinkException(
+				"Error while shutting down Kubernetes resource manager", firstException));
+		} else {
+			return terminationFuture;
+		}
```

Comment:
We could extract the below common codes with `YarnResourceManager` to parent `ResourceManager` if you like.

        ```
        final CompletableFuture<Void> terminationFuture = super.onStop();
		if (firstException != null) {
			return FutureUtils.completedExceptionally(new FlinkException(
				"Error while shutting down Kubernetes resource manager", firstException));
		} else {
			return terminationFuture;
		}
```

Reply:
I will add a new method `getStopTerminationFutureOrCompletedExceptionally` in `ResourceManager` for the common codes.